File: 767. Reorganize String.java
class Solution {
    class Pair{
        char ch;
        int count;
        public Pair(char ch, int count){
            this.ch = ch;
            this.count = count;
        }
    }
    public String reorganizeString(String s) {
        PriorityQueue<Pair> pq = new PriorityQueue<>((a, b) -> Integer.compare(b.count, a.count));
        HashMap<Character, Integer> hm = new HashMap<>();
        for(char ch : s.toCharArray()) hm.put(ch, hm.getOrDefault(ch, 0) + 1);

        for(Map.Entry<Character, Integer> ele : hm.entrySet()) pq.offer(new Pair(ele.getKey(), ele.getValue()));

        StringBuilder sb = new StringBuilder();
        while(!pq.isEmpty() && pq.size() >= 2){
            Pair first = pq.poll();
            Pair second = pq.poll();
            sb.append(first.ch).append(second.ch);
            if(first.count != 1) pq.offer(new Pair(first.ch, first.count - 1));
            if(second.count != 1) pq.offer(new Pair(second.ch, second.count - 1));
        }
        if(pq.size() == 1 && pq.peek().count == 1 && (sb.isEmpty() || sb.charAt(sb.length() - 1) != pq.peek().ch)) sb.append(pq.poll().ch);
        
        System.out.println(sb);
        if(!pq.isEmpty()) return "";
        return sb.toString();
    }
}

File: 767. Reorganize String.cpp
class Solution {
public:
    struct Pair {
        char ch;
        int count;
        Pair(char c, int cnt) : ch(c), count(cnt) {}
    };

    struct Compare {
        bool operator()(const Pair& a, const Pair& b) {
            return a.count < b.count; 
        }
    };

    string reorganizeString(string s) {
        priority_queue<Pair, vector<Pair>, Compare> pq;
        unordered_map<char, int> freq;

        for (char ch : s) freq[ch]++;

        for (auto& [ch, cnt] : freq) pq.push(Pair(ch, cnt));

        string result;

        while (!pq.empty() && pq.size() >= 2) {
            Pair first = pq.top(); pq.pop();
            Pair second = pq.top(); pq.pop();

            result += first.ch;
            result += second.ch;

            if (first.count > 1) pq.push(Pair(first.ch, first.count - 1));
            if (second.count > 1) pq.push(Pair(second.ch, second.count - 1));
        }

        if (!pq.empty()) {
            Pair last = pq.top();
            if (last.count == 1 && (result.empty() || result.back() != last.ch))
                result += last.ch;
            else
                return "";
        }

        return result;
    }
};